kornia/feature/laf.py

```diff
@@ -0,0 +1,381 @@
+            "LAF shape should be must be [Nx2x3] or [BxNx2x3]. "
+            "Got {}".format(LAF.size())
+        )
+    n, ch, h, w = images.size()
+    w = float(w)
+    h = float(h)
+    min_size = min(h, w)
+    coef = torch.ones(1, 1, 2, 3).to(LAF.dtype) * min_size
+    coef[0, 0, 0, 2] = w
+    coef[0, 0, 1, 2] = h
+    coef.to(LAF.device)
+    return coef.expand_as(LAF) * LAF
+
+
+def normalize_LAF(LAF: torch.Tensor, images: torch.Tensor) -> torch.Tensor:
+    """
+    Normalizes LAFs to [0,1] scale.
+    B,N,H,W = images.size()
+    MIN_SIZE = min(H,W)
+    [a11 a21 x]
+    [a21 a22 y]
+    becomes
+    [a11/MIN_SIZE a21/MIN_SIZE x/W]
+    [a21/MIN_SIZE a22/MIN_SIZE y/H]
+    Args:
+        LAF: (torch.Tensor).
+        images: (torch.Tensor) images, LAFs are detected in
+
+    Returns:
+        LAF: (torch.Tensor).
+
+    Shape:
+        - Input: :math:`(B, N, 2, 3)`
+        - Output:  :math:`(B, N, 2, 3)`
+    """
+    n_dims = len(LAF.size())
+    if (n_dims != 4):
+        raise TypeError(
+            "LAF shape should be must be [BxNx2x3]. "
+            "Got {}".format(LAF.size()))
+    n, ch, h, w = images.size()
+    w = float(w)
+    h = float(h)
+    min_size = min(h, w)
+    coef = torch.ones(1, 1, 2, 3).to(LAF.dtype) / min_size
+    coef[0, 0, 0, 2] = 1.0 / w
+    coef[0, 0, 1, 2] = 1.0 / h
+    coef.to(LAF.device)
+    return coef.expand_as(LAF) * LAF
+
+
+def generate_patch_grid_from_normalized_LAF(img: torch.Tensor,
+                                            LAF: torch.Tensor,
+                                            PS: int = 32) -> torch.Tensor:
+    """
+    Helper function for affine grid generation
+    """
+    n_dims = len(LAF.size())
+    if (n_dims != 4):
+        raise TypeError(
+            "LAF shape should be must be [BxNx2x3]. "
+            "Got {}".format(LAF.size()))
+    B, N, _, _ = LAF.size()
+    num, ch, h, w = img.size()
+    # norm, then renorm is needed for allowing detection on one resulution
+    # and extraction at arbitrary other
+    LAF_renorm = denormalize_LAF(LAF, img)
+
+    grid = F.affine_grid(LAF_renorm.view(B * N, 2, 3),
+                         [B * N, ch, PS, PS])
+    grid[:, :, :, 0] = 2.0 * grid[:, :, :, 0].clone() / float(w) - 1.0
+    grid[:, :, :, 1] = 2.0 * grid[:, :, :, 1].clone() / float(h) - 1.0
+    return grid
+
+
+def extract_patches_simple(img: torch.Tensor,
+                           LAF: torch.Tensor,
+                           PS: int = 32) -> torch.Tensor:
+    """
+    Extract patches defined by LAFs from image tensor.
+    No smoothing applied, huge aliasing (better use extract_patches_from_pyramid)
+    Args:
+        LAF: (torch.Tensor).
+        images: (torch.Tensor) images, LAFs are detected in
+        PS: (int) patch size, default = 32
+
+    Returns:
+        LAF: (torch.Tensor),  :math:`(B, N, CH, PS,PS)`
+    """
+    num, ch, h, w = img.size()
+    B, N, _, _ = LAF.size()
+    out = []
+    # for loop temporarily, to be refactored
```

Comment:
I think this can be done using `kornia.warp_perspective` or `kornia.warp_affine` correct me if I'm wrong

Reply:
I can, but for warp_affine there is a very different convention for affine matrix as an input. I feel, that it is more clear to have two separate implementations.

Reply:
okay, but we might want to converge to same code in the future to avoid bugs